### PR TITLE
Added support for approval when build is run

### PR DESCRIPTION
**Pull request by abhinavrau**
Fixes hashicorp/terraform-provider-google#9880

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
cloudbuild: Add approval config
```

### CODE REVIEW

**Comment by wyardley** [comment]
Woo! Much appreciated @abhinavrau

**Comment by abhinavrau** [comment]
Not sure why this test failed since my PR does not have anything to do with what this test is testing

**Comment by wyardley** [comment]
@abhinavrau best I can see, it's failing for other PRs, so maybe expected? Also, I think it looks like a maintainer may need to run `/gcbrun` or something to trigger additional cloudbuild tests?

**Comment by shuyama1** [comment]
Thanks @abhinavrau for making the changes and thanks @wyardley for looking into the issue. Test failed seem not related to this PR. I am working on reviewing the PR and testing it locally.

**Review by shuyama1 — CHANGES_REQUESTED** [review]
Tested the field locally and it seemed like setting `approval_required` to false will fail the test

**Comment by shuyama1** [comment]
The problems are
1. The value is not got sent if `approval_config.approval_required` is set to `false`
    This can be solve by adding `send_empty_value` to the `approval_required` field to force sending to value in the request body.
2. The API does return the value when it is set to `false`, which would then cause a perma-diff after we force sending the value.
    This could be solved by adding a custom flattener for `approval_config`

**Comment by abhinavrau** [comment]
Thanks @shuyama1 for finding the issue. When `approval_config.approval_required` is set to `false`  and then an `terraform apply` followed by a `terraform plan` is run, it shows changes which failing the test.

> The problems are
> 
> 1. The value is not got sent if `approval_config.approval_required` is set to `false`
>    This can be solve by adding `send_empty_value` to the `approval_required` field to force sending to value in the request body.


The above is not the case since the API does not expect it to be sent when `false`


> 2. The API does return the value when it is set to `false`, which would then cause a perma-diff after we force sending the value.
>    This could be solved by adding a custom flattener for `approval_config`

The above **is the problem** when this flag is set to `false`. I added a `flatten_object: true` in the `terraform.yaml` file with a custom flattener as you suggested. I based my flattener on this [one](https://github.com/GoogleCloudPlatform/magic-modules/blob/1d293f7bfadacaa20580874c8e8634827fb99a14/mmv1/templates/terraform/custom_flatten/health_check_log_config.go.erb) which should work. But I am facing the following issue when I run the `TestAccCloudBuildTrigger_cloudbuildTriggerManualExample` test. 

```2022/03/24 01:46:42 [WARN] Got error running Terraform: exit status 1

Error: Plugin did not respond

  with google_cloudbuild_trigger.manual-trigger,
  on terraform_plugin_test.tf line 3, in resource "google_cloudbuild_trigger" "manual-trigger":
   3: resource "google_cloudbuild_trigger" "manual-trigger" {

The plugin encountered an error, and failed to respond to the
plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain
more details.
    provider_test.go:301: Step 1/3 error: Error running apply: exit status 1
        
        Error: Plugin did not respond
        
          with google_cloudbuild_trigger.manual-trigger,
          on terraform_plugin_test.tf line 3, in resource "google_cloudbuild_trigger" "manual-trigger":
           3: resource "google_cloudbuild_trigger" "manual-trigger" {
        
        The plugin encountered an error, and failed to respond to the
        plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain
        more details.
```
The output does not tell me much of what went wrong. I tried adding `log.Printf` statements to the flatterner but it does not print them out. I know it gets to the flattener code since removing the code in the flattener with just the log statements provide some output. Any ideas on what could be wrong?

Also any tips on how I can debug the binary when the test is run?

**Comment by shuyama1** [comment]
> Thanks @shuyama1 for finding the issue. When `approval_config.approval_required` is set to `false` and then an `terraform apply` followed by a `terraform plan` is run, it shows changes which is failing the test.
> 
> > The problems are
> > 
> > 1. The value is not got sent if `approval_config.approval_required` is set to `false`
> >    This can be solve by adding `send_empty_value` to the `approval_required` field to force sending to value in the request body.
> 
> The above is not the case since the API does not expect it to be sent when `false`
> 
> > 2. The API does return the value when it is set to `false`, which would then cause a perma-diff after we force sending the value.
> >    This could be solved by adding a custom flattener for `approval_config`
> 
> The above **is the problem** when this flag is set to `false`. I added a `flatten_object: true` in the `terraform.yaml` file with a custom flattener as you suggested. I based my flattener on this [one](https://github.com/GoogleCloudPlatform/magic-modules/blob/1d293f7bfadacaa20580874c8e8634827fb99a14/mmv1/templates/terraform/custom_flatten/health_check_log_config.go.erb) which should work. But I am facing the following issue when I run the `TestAccCloudBuildTrigger_cloudbuildTriggerManualExample` test.
> 
> ```
> 
> Error: Plugin did not respond
> 
>   with google_cloudbuild_trigger.manual-trigger,
>   on terraform_plugin_test.tf line 3, in resource "google_cloudbuild_trigger" "manual-trigger":
>    3: resource "google_cloudbuild_trigger" "manual-trigger" {
> 
> The plugin encountered an error, and failed to respond to the
> plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain
> more details.
>     provider_test.go:301: Step 1/3 error: Error running apply: exit status 1
>         
>         Error: Plugin did not respond
>         
>           with google_cloudbuild_trigger.manual-trigger,
>           on terraform_plugin_test.tf line 3, in resource "google_cloudbuild_trigger" "manual-trigger":
>            3: resource "google_cloudbuild_trigger" "manual-trigger" {
>         
>         The plugin encountered an error, and failed to respond to the
>         plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain
>         more details.
> ```
> 
> The output does not tell me much of what went wrong. I tried adding `log.Printf` statements to the flatterner but it does not print them out. I know it gets to the flattener code since removing the code in the flattener with just the log statements provide some output. Any ideas on what could be wrong?

Would you mind committing your code?

**Comment by abhinavrau** [comment]
Never mind. I debugged the test and it turns out I had to use the fieldname as specified in the terraform.yaml to the flattened field name. It is working now.

File: mmv1/templates/terraform/custom_flatten/cloudbuild_approval_required.go.erb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+	if v == nil {
+		// Disabled by default, but API will not return object if value is false
+		transformed["approval_required"] = false
+		return []interface{}{transformed}
+	}
+
+	original := v.(map[string]interface{})
+	transformed["approval_required"] = original["approvalRequired"]
+	return []interface{}{transformed}
+}
```
Comment (wyardley) [inline mmv1/templates/terraform/custom_flatten/cloudbuild_approval_required.go.erb:26]: nit: not sure if matters for this project, but missing UNIX newline here.

Comment (abhinavrau) [reply mmv1/templates/terraform/custom_flatten/cloudbuild_approval_required.go.erb:26]: Sure. Will add it

**Comment by abhinavrau** [comment]
These failures are not related to this PR in any way. Could you please
review the results to see if this PR looks good?

Thanks

**Comment by shuyama1** [comment]
@abhinavrau Thanks for making the update and sorry for the delay on reviewing it! Got a question tho: any reason that we want to flatten the nested object here? I think we'll want to preserve the structure of the resource as it was, as we try to limit the use of changing the schema of resources.
Do you think the following plan would make sense in this case?
Instead of flatten `approval_config. approval_required` to the top, we can keep the original nested structure (remove the `flatten_object: true` but keep the `custom flatten`), add `default_from_api: true` to `approval_config` in terraform.yaml - it will mark `approval_config` as optional + computed, which means that not setting it in the config won't show a perma-diff in the terraform.

**Comment by abhinavrau** [comment]
Thanks @shuyama1 for reviewing and the suggestion. Let me try it out.

File: mmv1/products/cloudbuild/terraform.yaml [hunk 2]
```diff
@@ -77,6 +77,11 @@ overrides: !ruby/object:Overrides::ResourceOverrides
           One of `trigger_template`, `github`, `pubsub_config` `webhook_config` or `source_to_build` must be provided.
       triggerTemplate.projectId: !ruby/object:Overrides::Terraform::PropertyOverride
         default_from_api: true
+      approvalConfig: !ruby/object:Overrides::Terraform::PropertyOverride
+        default_from_api: true
+        custom_flatten: templates/terraform/custom_flatten/cloudbuild_approval_required.go.erb
+      approvalConfig.approvalRequired: !ruby/object:Overrides::Terraform::PropertyOverride
+        name: approval_required
```
Comment (shuyama1) [inline mmv1/products/cloudbuild/terraform.yaml:84]: Then we probably do not need the name override here

**Review by shuyama1 — CHANGES_REQUESTED** [review]

**Review by shuyama1 — APPROVED** [review]
LGTM! Thank you so much!

**Comment by wyardley** [comment]
Thanks @abhinavrau @shuyama1! This has been on my want list for a while! Looking forward to seeing this in an upcoming provider release.

**Pull request merged**